sw/device/lib/testing/test_framework/freertos_port.S

```diff
@@ -0,0 +1,163 @@
+
+  // Initialize t0 to the value of MSTATUS with global interrupts enabled, which
+  // is required because this returns with ret, not eret.
+  lw x5, 29 * portWORD_SIZE( sp )
+  addi x5, x5, 0x08
+  csrrw x0, mstatus, x5 // Ibex interrupts enabled from here!
+
+  // Restore t0 register from the stack (after using it to manipulate MSTATUS).
+  lw x5, 2 * portWORD_SIZE( sp )
+
+  // Upadate the stack pointer (shrinking the stack).
+  addi  sp, sp, portCONTEXT_SIZE
+
+  ret
+
+// -----------------------------------------------------------------------------
+
+// The prototype for this function depends on configurations defined in
+// FreeRTOSConfig.h, and is defined in:
+// sw/vendor/freertos_freertos_kernel/include/portable.h
+//
+// The implementation of this assembly function assumes the prototype for this
+// function looks like:
+//
+// StackType_t *pxPortInitialiseStack(StackType_t *pxTopOfStack,
+//                                    TaskFunction_t pxCode,
+//                                    void *pvParameters);
+//
+// TODO: add some checks to verify this is the configured prototype.
+// TODO: configure to allow use of checking for stack overflows.
+// TODO: configure return address for first (main) task.
+//
+// As per the standard RISC-V ABI pxTopcOfStack is passed in in a0, pxCode in
+// a1, and pvParameters in a2. The new top of stack is passed out in a0.
+//
+// The RISC-V context is saved to FreeRTOS tasks in the following stack frame,
+// where the global and thread pointers are currently assumed to be constant,
+// and therefore are not saved:
+//
+// ---Stack Bottom---
+// ---............---
+// Offset - Reg/Value
+//     29 - mstatus
+//     28 - t6 (x31)
+//     27 - t5 (x30)
+//     26 - t4 (x29)
+//     25 - t3 (x28)
+//     24 - s11 (x27)
+//     23 - s10 (x26)
+//     22 - s9 (x25)
+//     21 - s8 (x24)
+//     20 - s7 (x23)
+//     19 - s6 (x22)
+//     18 - s5 (x21)
+//     17 - s4 (x20)
+//     16 - s3 (x19)
+//     15 - s2 (x18)
+//     14 - a7 (x17)
+//     13 - a6 (x16)
+//     12 - a5 (x15)
+//     11 - a4 (x14)
+//     10 - a3 (x13)
+//      9 - a2 (x12)
+//      8 - a1 (x11)
+//      7 - (pvParameters)
+//      6 - s1 (x9)
+//      5 - s0 (x8)
+//      4 - t2 (x7)
+//      3 - t1 (x6)
+//      2 - t0 (x5)
+//      1 - (return address for main task, 0 for now)
+//      0 - (pxCode)
+// -----Stack Top----
+
+
+.balign 8
+pxPortInitialiseStack:
+
+  // Setup the MSTATUS register.
+  csrr t0, mstatus
+  // Ensure interrupts are disabled when the stack is restored within an ISR.
+  // Required when a task is created after the schedulre has been started,
+  // otherwise interrupts would be disabled anyway.
+  andi t0, t0, ~0x8
+  // Generate the value 0x1880, to set the MPIE and MPP bits in MSTATUS.
+  addi t1, x0, 0x188
+  slli t1, t1, 4
+  or t0, t0, t1
+
+  // Setup the stack frame detailed above (a0 holds the task stack pointer).
+  // Push MSTATUS onto the stack.
+  addi a0, a0, -portWORD_SIZE
```

Comment:
I don't think it actually matters, but what do you think of writing
```S
addi a0, a0 -29 * WORD_SIZE
sw a1, 0(a0)
sw zero, 1(a0)
// and so on
```

Reply:
Fixed.